metrics.go

```diff
@@ -50,7 +50,12 @@ func (c *sidekickCollector) Describe(ch chan<- *prometheus.Desc) {
 
 // Collect is called by the Prometheus registry when collecting metrics.
 func (c *sidekickCollector) Collect(ch chan<- prometheus.Metric) {
-	for _, c := range globalConnStats {
+	// automatically read the global stats
+	// Read/Write Lock is not required here
+	globalConnStatsRWMutex.RLock()
+	connStats := globalConnStats
+	globalConnStatsRWMutex.RUnlock()
```

Comment:
Since the array holds pointers, it just copies the pointers and the underlying object can still change. You're also writing to these objects, so you should have fetched a write-lock.

Reply:
Another (potential) issue here is that you first call `b.Stats.Lock()` and later you call `globalConnStatsRWMutex.RLock()`. If the `globalConnStatsRWMutex` lock blocks, then you also keep `b.Stats` locked. If you perform locks in a different order elsewhere in your code, then the code may deadlock.

These kind of issues are hard to find, so I tend to avoid that situation. I would rather fetch the required fields from `b.Stats` and then release the lock before locking `globalConnStatsRWMutex`.

Even if your code locks in the proper order right now, then someone else may add a lock somewhere and don't notice that it may cause a deadlock. That's why I am always very cautious with synchronization primitives (and rather avoid using locks whenever possible).

Reply:
Ah, I see. It is using atomic writes to set all values, so that's why a read-only lock is needed to fetch the connection status array. A read-only lock is fine, but there's still a potential issue that you lock different locks that may cause deadlocks. Using the `Pointer.Swap` can also be used for this type and you'll eliminate the lock.

PS: I also don't see any use of the `globalConnStats`. It's kept up-to-date, but I don't see where the value is actually used.

Reply:
It removes the lock, so that's already a benefit worth considering. And I don't see any drawbacks.

Reply:
Can you show me where the data inside `globalConnStats` is actually read? I couldn't find it.

Reply:
Ah, I see. But that could also call `Pointer.Load` to get access to a lock-less implementation.